# trial_visit_info 전처리 진행

- 3일체험 신청자 일자별 방문기록

In [22]:
import pandas as pd

df = pd.read_csv(r"C:\Users\nuwba\Desktop\Codeit_4\중급2_프로젝트\중급_2\trial_visit_info.csv")

df

site_id        date        stay_time  stay_time_second  \
0            1  2023-12-23  01:11:19.983486              4279   
1            1  2023-12-24  04:35:10.361341             16510   
2            1  2023-12-25  02:09:54.117163              7794   
3            1  2023-12-23  04:37:36.134901             16656   
4            3  2023-12-26  04:32:05.289251             16325   
...        ...         ...              ...               ...   
11472        6  2022-06-16  06:47:59.975925             24479   
11473        6  2022-06-17  01:28:46.221537              5326   
11474        6  2022-06-16  05:08:03.166662             18483   
11475        6  2023-02-06  06:30:29.693768             23429   
11476        6  2022-04-20  07:48:04.922610             28084   

                 first_enter_time             last_leave_time  \
0      2023-12-23 16:11:25.345793  2023-12-23 17:25:32.233995   
1      2023-12-24 16:10:45.668026  2023-12-24 20:53:34.842447   
2      2023-12-25 10:19:47.087268  2023-12-25 12:39:51.221372   
3      2023-12-23 14:49:40.302823  2023-12-23 19:27:39.406803   
4      2023-12-26 14:57:10.220793  2023-12-26 19:34:36.971542   
...                           ...                         ...   
11472  2022-06-16 11:00:48.643935  2022-06-16 18:23:38.495290   
11473  2022-06-17 17:15:05.221223  2022-06-17 18:56:00.840639   
11474  2022-06-16 12:52:46.557742  2022-06-16 18:23:34.355309   
11475  2023-02-06 10:14:00.538281  2023-02-06 16:52:37.606175   
11476  2022-04-20 13:24:06.222380         2022-04-20 23:59:59   

                                  user_uuid  
0      939fc566-61cd-47f0-a3ce-e2936fdcff73  
1      939fc566-61cd-47f0-a3ce-e2936fdcff73  
2      939fc566-61cd-47f0-a3ce-e2936fdcff73  
3      ba49324f-d317-4053-adf9-2b17b3f07594  
4      ba49324f-d317-4053-adf9-2b17b3f07594  
...                                     ...  
11472  766221e7-a753-444f-aa5a-ae76ce86d3fd  
11473  0a70277f-a7f9-401c-95fa-ed6d96ee32e0  
11474  0a70277f-a7f9-401c-95fa-ed6d96ee32e0  
11475  576473bb-0647-483e-b06a-a5ff6491b08b  
11476  ac90eb79-233a-47e0-99c7-c8f7b8372a57  

[11477 rows x 7 columns]

## 전처리 항목
- row 중에 `first_enter_time` / `last_leave_time` 둘 다 결측인 row
- `stay_time` 이 존재하고 `first_enter_time`, `last_leave_time` 둘 중 한 개만 존재하는 데이터
- `user_uuid` 3번 초과로 등장하는 데이터
- `stay_time`, `first_enter_time`, `last_leave_time` 가 존재하는데, 계산했을 때 `stay_time` 이안맞는 경우

### 1. 각 컬럼 내 결측값 확인

In [23]:
df.isna().sum()

site_id               0
date                  0
stay_time             0
stay_time_second      0
first_enter_time    555
last_leave_time     555
user_uuid             0
dtype: int64

`first_enter_time`  / `last_leave_time`  에 각 555건씩 결측 값 존재  

#### `stay_time` 기준 `first_enter_time`  / `last_leave_time` 두 속성이 동시에 결측인 row 수

In [24]:
# stay_time 존재 + first_enter_time / last_leave_time 둘 다 결측인 row 수
mask = df['stay_time'].notna() & df['first_enter_time'].isna() & df['last_leave_time'].isna()
count = mask.sum()

print(f"조건에 해당하는 row 수: {count}")

조건에 해당하는 row 수: 555


- `stay_time`은 존재하지만 `first_enter_time`과 `last_leave_time` 중 한쪽만 존재하는 데이터는 발견되지 않음
- 결측에 있어서 해당되는 555건은 삭제 진행

In [25]:
df.shape

(11477, 7)

In [26]:
# stay_time 존재 + first_enter_time/last_leave_time 둘 다 결측 → 삭제
mask = df['stay_time'].notna() & df['first_enter_time'].isna() & df['last_leave_time'].isna()
count = mask.sum()

df_cleaned = df[~mask].reset_index(drop=True)

print(f"삭제된 row 수: {count}")
print(f"삭제 후 데이터 shape: {df_cleaned.shape}")

삭제된 row 수: 555
삭제 후 데이터 shape: (10922, 7)


---

In [27]:
# 각 속성(열) 결측치 제거
df_cleaned

site_id        date        stay_time  stay_time_second  \
0            1  2023-12-23  01:11:19.983486              4279   
1            1  2023-12-24  04:35:10.361341             16510   
2            1  2023-12-25  02:09:54.117163              7794   
3            1  2023-12-23  04:37:36.134901             16656   
4            3  2023-12-26  04:32:05.289251             16325   
...        ...         ...              ...               ...   
10917        6  2022-06-16  06:47:59.975925             24479   
10918        6  2022-06-17  01:28:46.221537              5326   
10919        6  2022-06-16  05:08:03.166662             18483   
10920        6  2023-02-06  06:30:29.693768             23429   
10921        6  2022-04-20  07:48:04.922610             28084   

                 first_enter_time             last_leave_time  \
0      2023-12-23 16:11:25.345793  2023-12-23 17:25:32.233995   
1      2023-12-24 16:10:45.668026  2023-12-24 20:53:34.842447   
2      2023-12-25 10:19:47.087268  2023-12-25 12:39:51.221372   
3      2023-12-23 14:49:40.302823  2023-12-23 19:27:39.406803   
4      2023-12-26 14:57:10.220793  2023-12-26 19:34:36.971542   
...                           ...                         ...   
10917  2022-06-16 11:00:48.643935  2022-06-16 18:23:38.495290   
10918  2022-06-17 17:15:05.221223  2022-06-17 18:56:00.840639   
10919  2022-06-16 12:52:46.557742  2022-06-16 18:23:34.355309   
10920  2023-02-06 10:14:00.538281  2023-02-06 16:52:37.606175   
10921  2022-04-20 13:24:06.222380         2022-04-20 23:59:59   

                                  user_uuid  
0      939fc566-61cd-47f0-a3ce-e2936fdcff73  
1      939fc566-61cd-47f0-a3ce-e2936fdcff73  
2      939fc566-61cd-47f0-a3ce-e2936fdcff73  
3      ba49324f-d317-4053-adf9-2b17b3f07594  
4      ba49324f-d317-4053-adf9-2b17b3f07594  
...                                     ...  
10917  766221e7-a753-444f-aa5a-ae76ce86d3fd  
10918  0a70277f-a7f9-401c-95fa-ed6d96ee32e0  
10919  0a70277f-a7f9-401c-95fa-ed6d96ee32e0  
10920  576473bb-0647-483e-b06a-a5ff6491b08b  
10921  ac90eb79-233a-47e0-99c7-c8f7b8372a57  

[10922 rows x 7 columns]

### 2. `user_uuid` 3번 초과로 등장하는지 확인
- 3일체험 유저들의 일자별 방문기록 기준인지 확인

#### 체험일자별 전체 비율 확인

In [28]:
# user_uuid별 등장 횟수 집계
user_counts = df_cleaned['user_uuid'].value_counts()

# 그룹핑
one_day     = (user_counts == 1).sum()   # 1번만 등장
two_day     = (user_counts == 2).sum()   # 2번 등장
three_day   = (user_counts == 3).sum()   # 3번 등장
over_three  = (user_counts > 3).sum()    # 3번 초과

# 전체 유저 수
total_users = len(user_counts)

# 결과 정리
result = pd.DataFrame({
    "유저 수": [one_day, two_day, three_day, over_three],
    "비율(%)": [
        one_day / total_users * 100,
        two_day / total_users * 100,
        three_day / total_users * 100,
        over_three / total_users * 100
    ]
}, index=["1일 체험", "2일 체험", "3일 체험", "3일 초과"])

print(result.round(2))

       유저 수  비율(%)
1일 체험  3011  48.51
2일 체험  1831  29.50
3일 체험  1244  20.04
3일 초과   121   1.95


#### 전체 데이터 중 **3일 초과 유저들의 비율**

In [29]:
# user_uuid별 등장 횟수
user_counts = df_cleaned['user_uuid'].value_counts()

# 3번 초과 등장하는 유저 id 목록
over3_ids = user_counts[user_counts > 3].index

# 해당 유저들의 row만 추출
over3_rows = df_cleaned[df_cleaned['user_uuid'].isin(over3_ids)]

# 집계
num_users = len(over3_ids)                      # 유저 수
num_rows = len(over3_rows)                      # row 수
ratio_rows = num_rows / len(df_cleaned) * 100   # 전체 대비 %

print(f"3번 초과 유저 수: {num_users}")
print(f"이들이 차지하는 row 수: {num_rows}")
print(f"전체 데이터에서 차지 비율: {ratio_rows:.2f}%")

3번 초과 유저 수: 121
이들이 차지하는 row 수: 517
전체 데이터에서 차지 비율: 4.73%


- **3번 초과 유저를 이상치라 판단**
- `3일 체험 신청자`라는 기준이 있기 때문에
- n일차에 취소 후 다음 날을 사용했겠다 라는 가능성이 있었지만,
- `약 5%`정도의 데이터여서 삭제로 판단함

In [30]:
# 3번 초과 유저 id 추출
user_counts = df_cleaned['user_uuid'].value_counts()
over3_ids = user_counts[user_counts > 3].index

# df_cleaned에서 제외
df_cleaned_2 = df_cleaned[~df_cleaned['user_uuid'].isin(over3_ids)].reset_index(drop=True)
deleted_rows = df_cleaned['user_uuid'].isin(over3_ids).sum()

print(f"삭제된 유저 수: {len(over3_ids)}")
print(f"삭제된 행 수: {deleted_rows}")
print(f"삭제 후 데이터 shape: {df_cleaned_2.shape}")

삭제된 유저 수: 121
삭제된 행 수: 517
삭제 후 데이터 shape: (10405, 7)


#### 이상치 삭제 후 전체데이터 비율 확인

In [31]:
# user_uuid별 등장 횟수 집계
user_counts = df_cleaned_2['user_uuid'].value_counts()

# 그룹핑
one_day     = (user_counts == 1).sum()   # 1번만 등장
two_day     = (user_counts == 2).sum()   # 2번 등장
three_day   = (user_counts == 3).sum()   # 3번 등장
over_three  = (user_counts > 3).sum()    # 3번 초과

# 전체 유저 수
total_users = len(user_counts)

# 결과 정리
result = pd.DataFrame({
    "유저 수": [one_day, two_day, three_day, over_three],
    "비율(%)": [
        one_day / total_users * 100,
        two_day / total_users * 100,
        three_day / total_users * 100,
        over_three / total_users * 100
    ]
}, index=["1일 체험", "2일 체험", "3일 체험", "3일 초과"])

print(result.round(2))

       유저 수  비율(%)
1일 체험  3011  49.47
2일 체험  1831  30.09
3일 체험  1244  20.44
3일 초과     0   0.00


---

In [32]:
df_cleaned_2

site_id        date        stay_time  stay_time_second  \
0            1  2023-12-23  01:11:19.983486              4279   
1            1  2023-12-24  04:35:10.361341             16510   
2            1  2023-12-25  02:09:54.117163              7794   
3            1  2023-12-23  04:37:36.134901             16656   
4            3  2023-12-26  04:32:05.289251             16325   
...        ...         ...              ...               ...   
10400        6  2022-06-16  06:47:59.975925             24479   
10401        6  2022-06-17  01:28:46.221537              5326   
10402        6  2022-06-16  05:08:03.166662             18483   
10403        6  2023-02-06  06:30:29.693768             23429   
10404        6  2022-04-20  07:48:04.922610             28084   

                 first_enter_time             last_leave_time  \
0      2023-12-23 16:11:25.345793  2023-12-23 17:25:32.233995   
1      2023-12-24 16:10:45.668026  2023-12-24 20:53:34.842447   
2      2023-12-25 10:19:47.087268  2023-12-25 12:39:51.221372   
3      2023-12-23 14:49:40.302823  2023-12-23 19:27:39.406803   
4      2023-12-26 14:57:10.220793  2023-12-26 19:34:36.971542   
...                           ...                         ...   
10400  2022-06-16 11:00:48.643935  2022-06-16 18:23:38.495290   
10401  2022-06-17 17:15:05.221223  2022-06-17 18:56:00.840639   
10402  2022-06-16 12:52:46.557742  2022-06-16 18:23:34.355309   
10403  2023-02-06 10:14:00.538281  2023-02-06 16:52:37.606175   
10404  2022-04-20 13:24:06.222380         2022-04-20 23:59:59   

                                  user_uuid  
0      939fc566-61cd-47f0-a3ce-e2936fdcff73  
1      939fc566-61cd-47f0-a3ce-e2936fdcff73  
2      939fc566-61cd-47f0-a3ce-e2936fdcff73  
3      ba49324f-d317-4053-adf9-2b17b3f07594  
4      ba49324f-d317-4053-adf9-2b17b3f07594  
...                                     ...  
10400  766221e7-a753-444f-aa5a-ae76ce86d3fd  
10401  0a70277f-a7f9-401c-95fa-ed6d96ee32e0  
10402  0a70277f-a7f9-401c-95fa-ed6d96ee32e0  
10403  576473bb-0647-483e-b06a-a5ff6491b08b  
10404  ac90eb79-233a-47e0-99c7-c8f7b8372a57  

[10405 rows x 7 columns]

### 3. `first_enter_time`, `last_leave_time` 계산 했을 때 `stay_time`이 안맞는 경우

#### 계산이 안 맞는 행의 수와 비율

In [33]:
# 1) datetime 변환 (안 되어 있으면)
df_cleaned_2['first_enter_time'] = pd.to_datetime(df_cleaned_2['first_enter_time'], errors='coerce')
df_cleaned_2['last_leave_time'] = pd.to_datetime(df_cleaned_2['last_leave_time'], errors='coerce')

# 2) stay_time 문자열을 timedelta로 변환
df_cleaned_2['stay_time_td'] = pd.to_timedelta(df_cleaned_2['stay_time'], errors='coerce')

# 3) first ~ last 차이 구하기
df_cleaned_2['calculated_stay'] = df_cleaned_2['last_leave_time'] - df_cleaned_2['first_enter_time']

# 4) 불일치 여부 (허용 오차 ±1분)
mask = (df_cleaned_2['stay_time_td'].notna()) & (df_cleaned_2['calculated_stay'].notna()) & (
    (df_cleaned_2['stay_time_td'] - df_cleaned_2['calculated_stay']).abs() > pd.Timedelta(minutes=1)
)

# 5) 결과 확인
mismatch_count = mask.sum()
total_rows = len(df_cleaned_2)
mismatch_ratio = mismatch_count / total_rows * 100

print(f"불일치 행 수: {mismatch_count}")
print(f"전체 데이터 대비 비율: {mismatch_ratio:.2f}%")

# 필요시 불일치 행만 따로 보기
df_mismatch = df_cleaned_2[mask][['user_uuid','stay_time','stay_time_td','first_enter_time','last_leave_time','calculated_stay']]

불일치 행 수: 6639
전체 데이터 대비 비율: 63.81%


#### `first_enter_time` / `last_leave_time` 재 계산 후 `stay_time` 채우기

In [34]:
# 1) 복사본 생성
df_cleaned_3 = df_cleaned_2.copy()

# 2) stay_time 문자열 → timedelta 변환
df_cleaned_3['stay_time_td'] = pd.to_timedelta(df_cleaned_3['stay_time'], errors='coerce')

# 3) first ~ last 차이 계산
df_cleaned_3['calculated_stay'] = df_cleaned_3['last_leave_time'] - df_cleaned_3['first_enter_time']

# 4) 불일치 행 찾기 (±1분 이상 차이 나는 경우)
mask = (df_cleaned_3['stay_time_td'].notna()) & (df_cleaned_3['calculated_stay'].notna()) & (
    (df_cleaned_3['stay_time_td'] - df_cleaned_3['calculated_stay']).abs() > pd.Timedelta(minutes=1)
)

print(f"수정할 row 수: {mask.sum()}")

# 5) 불일치한 행의 stay_time을 재계산 값으로 교체
df_cleaned_3.loc[mask, 'stay_time'] = df_cleaned_3.loc[mask, 'calculated_stay'].astype(str)

# 6) 보조 컬럼 삭제
df_cleaned_3 = df_cleaned_3.drop(columns=['stay_time_td', 'calculated_stay'])

수정할 row 수: 6639


#### 처리한 `df_cleaned_3` 에서 불일치 항목 확인

In [35]:
# 1) stay_time 문자열 → timedelta 변환
df_cleaned_3['stay_time_td'] = pd.to_timedelta(df_cleaned_3['stay_time'], errors='coerce')

# 2) first ~ last 차이 계산
df_cleaned_3['calculated_stay'] = df_cleaned_3['last_leave_time'] - df_cleaned_3['first_enter_time']

# 3) 불일치 여부 체크 (±1분 이상 차이 나는 경우)
mask = (df_cleaned_3['stay_time_td'].notna()) & (df_cleaned_3['calculated_stay'].notna()) & (
    (df_cleaned_3['stay_time_td'] - df_cleaned_3['calculated_stay']).abs() > pd.Timedelta(minutes=1)
)

# 4) 결과 출력
mismatch_count = mask.sum()
total_rows = len(df_cleaned_3)
mismatch_ratio = mismatch_count / total_rows * 100

print(f"불일치 행 수: {mismatch_count}")
print(f"전체 데이터 대비 비율: {mismatch_ratio:.2f}%")

# 5) 실제 불일치 데이터 확인 (원하면)
df_mismatch = df_cleaned_3[mask][['user_uuid','first_enter_time','last_leave_time','stay_time']]

불일치 행 수: 0
전체 데이터 대비 비율: 0.00%


In [40]:
# stay_time을 timedelta로 변환 후 문자열로 변환 (마이크로초 포함)
df_cleaned_3['stay_time'] = pd.to_timedelta(df_cleaned_3['stay_time'], errors='coerce') \
                                .astype(str) \
                                .str.split(' ').str[-1]  # "0 days" 제거

df_cleaned_3 = df_cleaned_3.drop(['stay_time_td', 'calculated_stay'], axis=1)

In [41]:
df_cleaned_3

site_id        date        stay_time  stay_time_second  \
0            1  2023-12-23  01:14:06.888202              4279   
1            1  2023-12-24  04:42:49.174421             16510   
2            1  2023-12-25  02:20:04.134104              7794   
3            1  2023-12-23  04:37:36.134901             16656   
4            3  2023-12-26  04:37:26.750749             16325   
...        ...         ...              ...               ...   
10400        6  2022-06-16  07:22:49.851355             24479   
10401        6  2022-06-17  01:40:55.619416              5326   
10402        6  2022-06-16  05:30:47.797567             18483   
10403        6  2023-02-06  06:38:37.067894             23429   
10404        6  2022-04-20  07:48:04.922610             28084   

                first_enter_time            last_leave_time  \
0     2023-12-23 16:11:25.345793 2023-12-23 17:25:32.233995   
1     2023-12-24 16:10:45.668026 2023-12-24 20:53:34.842447   
2     2023-12-25 10:19:47.087268 2023-12-25 12:39:51.221372   
3     2023-12-23 14:49:40.302823 2023-12-23 19:27:39.406803   
4     2023-12-26 14:57:10.220793 2023-12-26 19:34:36.971542   
...                          ...                        ...   
10400 2022-06-16 11:00:48.643935 2022-06-16 18:23:38.495290   
10401 2022-06-17 17:15:05.221223 2022-06-17 18:56:00.840639   
10402 2022-06-16 12:52:46.557742 2022-06-16 18:23:34.355309   
10403 2023-02-06 10:14:00.538281 2023-02-06 16:52:37.606175   
10404 2022-04-20 13:24:06.222380                        NaT   

                                  user_uuid  
0      939fc566-61cd-47f0-a3ce-e2936fdcff73  
1      939fc566-61cd-47f0-a3ce-e2936fdcff73  
2      939fc566-61cd-47f0-a3ce-e2936fdcff73  
3      ba49324f-d317-4053-adf9-2b17b3f07594  
4      ba49324f-d317-4053-adf9-2b17b3f07594  
...                                     ...  
10400  766221e7-a753-444f-aa5a-ae76ce86d3fd  
10401  0a70277f-a7f9-401c-95fa-ed6d96ee32e0  
10402  0a70277f-a7f9-401c-95fa-ed6d96ee32e0  
10403  576473bb-0647-483e-b06a-a5ff6491b08b  
10404  ac90eb79-233a-47e0-99c7-c8f7b8372a57  

[10405 rows x 7 columns]

## csv 파일 생성

In [42]:
# CSV 저장 (UTF-8 인코딩)
df_cleaned_3.to_csv("trial_visit_info_cleaned.csv", index=False, encoding="utf-8-sig")